devito/passes/iet/engine.py

```diff
@@ -94,19 +96,50 @@ def apply(self, func, **kwargs):
             # we must update the call sites as well, as the arguments dropped down
             # to the efunc have just increased
             args = as_tuple(metadata.get('args'))
-            if args:
-                # `extif` avoids redundant updates to the parameters list, due
-                # to multiple children wanting to add the same input argument
-                extif = lambda v: list(v) + [e for e in args if e not in v]
-                stack = [i] + dag.all_downstreams(i)
-                for n in stack:
-                    efunc = self.efuncs[n]
-                    calls = [c for c in FindNodes(Call).visit(efunc) if c.name in stack]
-                    mapper = {c: c._rebuild(arguments=extif(c.arguments)) for c in calls}
-                    efunc = Transformer(mapper).visit(efunc)
-                    if efunc.is_Callable:
-                        efunc = efunc._rebuild(parameters=extif(efunc.parameters))
-                    self.efuncs[n] = efunc
+            if not args:
+                continue
+
+            def filter_args(v, efunc=None):
+                processed = list(v)
+                for _a in args:
+                    try:
+                        # Should the arg actually be dropped?
+                        a, drop = _a
+                        if drop:
```

Comment:
```
try:
    if a[1] and a[0] in processed:
        processed.remove(a[0])
    continue
except AttributeError:
    a = _a 
```

Reply:
I don't like explicit indexing into `a` , and I have a feeling that what you wrote is slightly different (semantically) than what I need here